### PR TITLE
Add flash for vtable destination, make it default, and add build menu to control options

### PR DESCRIPTION
**Pull request by earlephilhower**
Add an option for placing vtables in flash to complement the existing iram and heap options, and allow selection using a build menu.  No makefile needed any more,.

Extension of #4551 issue and PR #4567

### CODE REVIEW

**Comment by Adam5Wu** [comment]
@earlephilhower  I am not nitpicking on you or this PR. :)

In my recent exploration solving "__c causes a section type conflict with __c" (#3369) I found it is possible to define multiple sections for string literals, so that data with different tags can coexist.

Then I just wonders, whether it is possible to do similar things with vtable data? That is:
1. Put all vtable entries that are involved in ISR in one section, say `.vtable.isr`
2. Put all other vtable entries in another section, say `.vtable.regular`
3. Let the linker put `.vtable.isr` section in IRAM, and put `.vtable.regular` in flash

Just a crazy idea... not sure whether it is doable without hacking into the compiler.

**Comment by earlephilhower** [comment]
Unfortunately the compiler doesn't know which are interrupt called functions. I'm not sure you could tell that with static graph analysis, in fact.  Turing strikes again!

**Comment by Adam5Wu** [comment]
But could human annotation help?

Just like I know I am writing a template class, so instead of PSTR("Blah"), I use PSTR_T("Blah") to put my string literal to a different section to avoid conflict.

I know I am writing a class for ISR, then I add some annotation to the class so compiler can place the vtable of this class to the `.vtable.isr` section.

If that is possible I am happy enough...

**Comment by earlephilhower** [comment]
I don't think there's a way to annotate other than by class names.  You can individually specify in the .ld file to include, say, `_ZL*__irq`, in `.RODATA` (and hence in RAM), but I don't think I've ever seen any linker annotations on anything other than *instances* which do not have their own vtables.

If you find a way, we can figure out how to mangle it in, for sure!  I think in practice, though, class functions in IRQs are pretty rare, and virtual tables in those are even less prevalent.

**Comment by devyte** [comment]
Is there a way to be able to select this from the IDE menu? I'm thinking of the windows users.

**Comment by earlephilhower** [comment]
I think this can be in the menus if we change platform.txt and boards.py to build the app.ld file before every compile, like got GIT version.

In boards.txt there'll be (yet another) popup menu, identical for all devices, that adds a -D switch.

In platform.txt, add a recipe.hooks.prebuild.x to run the command the makefile does now.

I can give it a whirl, but it'll take a few days.

@d-a-v Does this make sense?  You've been hacking the build.py and platform.txt much more than I have...

**Comment by d-a-v** [comment]
Yes this could work this way and in fact would be much better because it'd work under windows without having to install make.

**Comment by devyte** [comment]
@d-a-v that's exactly what I was getting at :)

**Comment by earlephilhower** [comment]
@d-a-v , @devyte  I've just pushed a change which does what was discussed.  A new menu is used instead of a makefile, and no tools are required other than GCC so it should run fine under Windows, too.

**Comment by earlephilhower** [comment]
@d-a-v I'm stuck on how to make build.py get the default vtable flag that was added.  Interactive builds seem to just work and take the first item by default, but the build.py script is barfing about it being undefined:
````
"/home/travis/arduino_ide/hardware/esp8266com/esp8266/tools/xtensa-lx106-elf/bin/xtensa-lx106-elf-gcc" -CC -E -P {build.vtable_flags} "/home/travis/arduino_ide/hardware/esp8266com/esp8266/tools/sdk/ld/eagle.app.v6.common.ld.h" -o "/home/travis/arduino_ide/hardware/esp8266com/esp8266/tools/sdk/ld/eagle.app.v6.common.ld"
xtensa-lx106-elf-gcc: error: {build.vtable_flags}: No such file or directory
````

**Comment by d-a-v** [comment]
Setting the default value in platform.txt seems to solve it:
```
build.vtable_flags=-DVTABLES_IN_FLASH
```
It is overridden by the IDE.
(just verified both)

**Comment by earlephilhower** [comment]
Thanks, @d-a-v .  Looks to be building now!

File: platform.txt [hunk 1]
```diff
@@ -76,6 +78,9 @@ recipe.hooks.core.prebuild.2.pattern=bash -c "mkdir -p {build.path}/core && echo
 recipe.hooks.core.prebuild.1.pattern.windows=cmd.exe /c mkdir {build.path}\core & (echo #define ARDUINO_ESP8266_GIT_VER 0x00000000 & echo #define ARDUINO_ESP8266_GIT_DESC win-{version} ) > {build.path}\core\core_version.h
 recipe.hooks.core.prebuild.2.pattern.windows=
 
+## Build the app.ld linker file
+recipe.hooks.linking.prelink.1.pattern="{compiler.path}{compiler.c.cmd}" -CC -E -P {build.vtable_flags} "{runtime.platform.path}/tools/sdk/ld/eagle.app.v6.common.ld.h" -o "{runtime.platform.path}/tools/sdk/ld/eagle.app.v6.common.ld"
```
Comment (igrr) [inline platform.txt]: `eagle.app.v6.common.ld` is now auto-generated, why is it still tracked in Git? I think it should be removed from the repo and added to .gitignore.

**Comment by innodron** [comment]
I have  just pulled in the most recent commit ([e3c97021](https://github.com/esp8266/Arduino/commit/e3c970210fe74d35049afd0a7dc06759d4f5c437)) yesterday and now I am getting `xtensa-lx106-elf/bin/ld: cannot open linker script file ../ld/eagle.app.v6.common.ld: No such file or directory`error when compiling. Apparently this file is now removed from the repo and expected to be produced during a sketch build.

However, It seems `eagle.app.v6.common.ld` is not being auto-generated for some reason in my case. I  will appreciate to get guidance on how to approach to this issue. Apperently, no similar problems reported yet.

I am using Eclipse with Sloeber plugin. This is an ESP-12F module with the  [attached](https://user-images.githubusercontent.com/1423196/40218727-16350fd6-5a7b-11e8-8cdb-c93c5ccd4b90.png)  properties. Frankly I am not sure which value the new VTables option should take; I just assigned the default.

**Pull request merged**